File: e2e/tests/pkce.spec.ts
import { test, expect } from '@playwright/test';
import { setup, startAuthorization, runAuthCodeFlow } from './helpers';

test.describe('PKCE (S256)', () => {
  test('exchanges the code when the verifier matches', async ({ page }) => {
    const s = await setup(page, 'pkce');
    await runAuthCodeFlow(page, s, { pkce: true });
    await expect(page.locator('#exchange_result')).toContainText('"access_token"');
  });

  test('rejects the exchange with a missing or wrong verifier', async ({ page }) => {
    const s = await setup(page, 'pkce-neg');
    await startAuthorization(page, s, { pkce: true });
    await expect(page.locator('#authorize_result')).toContainText('"code"');

    // Missing verifier → invalid_request (missing required parameter).
    await page.locator('#tok_code_verifier').clear();
    await page.getByRole('button', { name: 'Exchange' }).click();
    await expect(page.locator('#exchange_result')).toContainText('invalid_request');

    // Wrong verifier → invalid_grant (challenge mismatch).
    await page.locator('#tok_code_verifier').fill('wrong-verifier-wrong-verifier-wrong-verifier');
    await page.getByRole('button', { name: 'Exchange' }).click();
    await expect(page.locator('#exchange_result')).toContainText('invalid_grant');
  });
});

File: spec/e2e/pkce_spec.rb
# frozen_string_literal: true

require_relative "e2e_helper"

RSpec.describe "PKCE (S256)" do
  it "exchanges the code when the verifier matches" do
    client = dashboard_client("pkce")
    run_auth_code_flow(client, pkce: true)

    expect(page).to have_css("#exchange_result", text: '"access_token"')
  end

  it "rejects the exchange with a missing or wrong verifier" do
    client = dashboard_client("pkce-neg")
    start_authorization(client, pkce: true)
    expect(page).to have_css("#authorize_result", text: '"code"')

    # Missing verifier → invalid_request (a required parameter is absent).
    fill_in "tok_code_verifier", with: ""
    click_button "Exchange"
    expect(page).to have_css("#exchange_result", text: "invalid_request")

    # Wrong verifier → invalid_grant (it does not match the challenge).
    fill_in "tok_code_verifier", with: "wrong-verifier-wrong-verifier-wrong-verifier"
    click_button "Exchange"
    expect(page).to have_css("#exchange_result", text: "invalid_grant")
  end
end
